KAFKA-14533: Do not interrupt state-updater thread during shutdown

guozhangwang opened the pull request:
1. Fix the StateUpdater shutdown procedure: a) in shutdown, we first set the running flag, then notify the condition; 2) in the thread's `waitIfAllChangelogsCompletelyRead` block, we collapse the `if` condition together with the `while` condition so that we always check all four conditions once the thread is notified inside the while loop. As a result, shutdown procedure would not involve any thread interruptions anymore.
2. Print fine-grained streams exception when list-offset fails, this is a byproduct of the debugging procedure but I think it's worth keeping since it has better operational visibilities.
3. Some nit logging improvements (including moving `logger` from the inner thread into the outer class to also add some more logging).
4. Re-enable state-updater in SmokeTestDriverIntegrationTest

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

guozhangwang committed:
wrap exception as streams exception

guozhangwang committed:
checkstyle

guozhangwang commented:
Re-starting new checks as `SmokeTestDriverIntegrationTest` did not fail again.

guozhangwang committed:
do not interrupt thread ever

guozhangwang committed:
cleanup

guozhangwang committed:
remove imports

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java:
```diff
@@ -143,6 +145,33 @@ public static KafkaFuture<Map<TopicPartition, ListOffsetsResultInfo>> fetchEndOf
         ).all();
     }
 
+    public static ListOffsetsResult fetchEndOffsetsResult(final Collection<TopicPartition> partitions,
```
This is 2) in the description.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -685,14 +685,17 @@ private boolean populateClientStatesMap(final Map<UUID, ClientState> clientState
-            // asynchronously while we use the blocking Consumer#committed call to fetch source-changelog offsets
-            final KafkaFuture<Map<TopicPartition, ListOffsetsResultInfo>> endOffsetsFuture =
-                fetchEndOffsetsFuture(changelogTopics.preExistingNonSourceTopicBasedPartitions(), adminClient);
+            // asynchronously while we use the blocking Consumer#committed call to fetch source-changelog offsets;
```
This is 2) in the description.

guozhangwang reviewed (commented):
Ping @lucasbru @cadonna for reviews.

Some context here: I found from the failed test that, the shutdown procedure with the `MAX` timeout never completes, as the thread was still waiting in the while loop inside `waitIfAllChangelogsCompletelyRead` function. Though I have not figured out exactly why the interrupt was missing --- one theory in my mind is that, by the time the thread was `interrupt()`, it was just created but `start()` not triggered yet --- in general I felt that relying on interrupt to shutdown the thread is not very ideal. So I tried to remove the interrupt mechanism but just rely on notifying the condition and cleaning up the `waitIfAllChangelogsCompletelyRead` while loop a bit. 

PS: I also thought about whether to stop hard-coding a `MAX.value` timeout in state updater shutdown, but I cannot think of a very good value here. We could pass in the timeout value from the outermost `KafkaStreams.close()` to it but if it is very small or even to zero, we may leave with a dangling state updater thread. So I decided to still keep this value.

Will see if jenkins jobs do not show this test failure any more (previously it fails 1 out of 3 times).

guozhangwang commented:
`SmokeTestDriverIntegrationTest` did not fail in `https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13318/2`, triggering again.

guozhangwang commented:
SmokeTestDriverIntegrationTest did not fail in https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13318/11, triggering again.

If that run still did not fail SmokeTestDriverIntegrationTest, I think we could be confident here. cc @lucasbru

lucasbru commented:
We had a failure of `Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.PauseResumeIntegrationTest.[1] true`. I'm pretty sure that one was not flaky before. Any chance this PR breaks it?

guozhangwang commented:
> We had a failure of Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.PauseResumeIntegrationTest.[1] true. I'm pretty sure that one was not flaky before. Any chance this PR breaks it? Also there are more failures due to the logging changes.

I fixed the test utils for all the `StreamsPartitionAssignorTest`; but I cannot find the failure of `PauseResumeIntegrationTest` in run 15, which run did you see it?

guozhangwang committed:
fix unit tests

guozhangwang committed:
checkstyle

lucasbru commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java:
```diff
@@ -312,16 +305,20 @@ private void addToExceptionsAndFailedTasksThenClearUpdatingTasks(final Exception
-                } finally {
-                    tasksAndActionsLock.unlock();
+        private void waitIfAllChangelogsCompletelyRead() {
+            tasksAndActionsLock.lock();
```
Nit: the name `tasksAndActionsLock/Condition` is probably not very adequate anymore. But we don't need to change it in this PR.

cadonna replied:
I agree! Actually, I would change it in this PR since it improves the readability of the code.

guozhangwang replied:
The name comes from the `TaskAndAction` class, which I think is still appropriate even if we are left with `add` and `remove` actions. What's your suggestion for a better naming?

cadonna replied:
I re-read the code and could not came up with a better name. Let's keep it as it is for now.

lucasbru reviewed (approved):
Thanks, I didn't realize that the failing `PauseResumeIntegrationTest` was from an earlier version that was functionally different. Assuming we don't get any new integration test failures, this lgtm

guozhangwang committed:
fix unit tests

guozhangwang commented:
@lucasbru did you happen to have triggered a system test for this branch as well?

lucasbru commented:
I triggered one now: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/5554/console

guozhangwang committed:
fix unit tests

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/RackAwarenessStreamsPartitionAssignorTest.java:
```diff
@@ -169,26 +166,6 @@ private void createMockTaskManager() {
         topologyMetadata.buildAndRewriteTopology();
     }
 
-    // If you don't care about setting the end offsets for each specific topic partition, the helper method
```
This is piggy-backed as part of incorporating 2): I found that these funcs are duplicated across multiple test classes from `AssignmentTestUtils.createMockAdminClientForAssignor`, so I removed them.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/HighAvailabilityStreamsPartitionAssignorTest.java:
```diff
@@ -149,35 +145,12 @@ private void configurePartitionAssignorWith(final Map<String, Object> props) {
         topologyMetadata.buildAndRewriteTopology();
     }
 
-    // If you don't care about setting the end offsets for each specific topic partition, the helper method
```
Ditto here.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/HighAvailabilityStreamsPartitionAssignorTest.java:
```diff
@@ -149,35 +145,12 @@ private void configurePartitionAssignorWith(final Map<String, Object> props) {
         overwriteInternalTopicManagerWithMock();
     }
 
-    // Useful for tests that don't care about the task offset sums
```
This is found that after we refactored with mockito, the passed in task set is not needed any more, so I removed them to eliminate the warnings.

guozhangwang reviewed (commented):
ping @mjsax for reviews.

guozhangwang commented:
ping @cadonna for a final look, this PR needs to go first before the next one.

guozhangwang commented:
> I triggered one now: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/5555/console

@lucasbru the failure seems not related to the code but sock timeouts itself. Have you seen similar scenarios before when working with system tests? Are they okay to ignore or not?

lucasbru commented:
Yeah, these seem to be unrelated. I restarted the job: 

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/5562/consoleText

We have two failures: 
test_id:    
```
kafkatest.tests.streams.streams_broker_bounce_test.StreamsBrokerBounceTest.test_many_brokers_bounce.failure_mode=clean_bounce.num_failures=2.metadata_quorum=ZK
kafkatest.tests.streams.streams_upgrade_test.StreamsUpgradeTest.test_rolling_upgrade_with_2_bounces.from_version=2.6.2.to_version=3.5.0-SNAPSHOT
```

It seems to me that they are unrelated.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java:
```diff
@@ -140,16 +134,15 @@ public void run() {
+                    // in state updater we never interrupt the thread
+                    // or wakeup the consumer, hence any of
+                    // InterruptException / InterruptedException / WakeupException
+                    // should never happen
```
We can get rid of this comment since `runOnce()` does not throw a `InterruptedException` anymore.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java:
```diff
@@ -564,14 +566,26 @@ public void start() {
             // initialize the last commit as of now to prevent first commit happens immediately
             this.lastCommitMs = time.milliseconds();
+
+            log.info("StateUpdater thread started");
```
I would let the state updater thread write this log message, not the thread that starts the state updater thread.

cadonna replied:
I just realized that we have the same log message on line 134. Can we remove this one? Should we also move the logger back to the thread?

guozhangwang replied:
Yeah I agree with you that the start up log line is a duplicate; at the same time I found that the log line for shutting down the thread from the caller thread is still valuable in debugging. So I would remove this line while keeping the logger still at the class level for that log line.

cadonna replied:
I see. What about having the log line 
```
Starting StateUpdater thread
```
here, then
```
StateUpdater thread started
```
in the beginning of method `run()` of the thread as well as
```
Shutting down StateUpdater thread
```
in `shutdown()` and
```
StateUpdater thread shut down
```
at the end of `run()` in the thread.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java:
```diff
@@ -312,16 +305,20 @@ private void addToExceptionsAndFailedTasksThenClearUpdatingTasks(final Exception
+                    tasksAndActions.isEmpty() &&
+                    !isTopologyResumed.get()) {
+                    tasksAndActionsCondition.await();
                 }
```
nit:
```suggestion
                while (isRunning.get() &&
                    changelogReader.allChangelogsCompleted() &&
                    tasksAndActions.isEmpty() &&
                    !isTopologyResumed.get()) {
                    
                    tasksAndActionsCondition.await();
                }
```

cadonna reviewed (commented):
Thanks for the PR. @guozhangwang !

Here my comments!

guozhangwang commented:
@cadonna Thanks for your comments. I've addressed your comments except the nit name.

LMK what do you think as a better name for the class `TaskAndAction` and I can change the class name, and other field names accordingly.

guozhangwang committed:
github comments

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java:
```diff
@@ -140,16 +134,15 @@ public void run() {
-                    } catch (final InterruptedException | InterruptException interruptedException) {
-                        return;
-                    }
+                    // in state updater we never interrupt the thread
+                    // or wakeup the consumer, hence any of
+                    // InterruptException / WakeupException
+                    // should never happen
+                    runOnce();
```
I still would like to get rid of this comment since `runOnce()` does not throw a `InterruptedException anymore`

guozhangwang replied:
ack, will remove upon merging

cadonna reviewed (approved):
LGTM!

Thanks @guozhangwang !

guozhangwang committed:
remove comment
